api/v1alpha1/shared_types.go

```diff
@@ -0,0 +1,162 @@
+}
+
+type GitLocation struct {
+	// RepoUrl is the URL to the repository itself
+	RepoUrl string `json:"repoUrl" mapstructure:"repoUrl"`
+
+	// Path is the full path from the root of the repository to where the resources are held
+	//  If the Path is empty, then the root directory will be used.
+	// Can be a file or a directory
+	// Note that all resources within this path (described by .yaml files) will be synced
+	Path string `json:"path" mapstructure:"path"`
+
+	// TargetRevision specifies the target revision to sync to, it can be a branch, a tag,
+	// or a commit hash.
+	TargetRevision string `json:"targetRevision" mapstructure:"targetRevision"`
+}
+
+// KustomizeSource holds kustomize specific options
+type KustomizeSource struct{}
+
+// HelmSource holds helm-specific options
+type HelmSource struct {
+	// ValuesFiles is a list of Helm value files to use when generating a template
+	ValueFiles []string `json:"valueFiles,omitempty" mapstructure:"valueFiles,omitempty"`
+	// Parameters is a list of Helm parameters which are passed to the helm template command upon manifest generation
+	Parameters []HelmParameter `json:"parameters,omitempty" mapstructure:"parameters,omitempty"`
+}
+
+// HelmParameter is a parameter passed to helm template during manifest generation
+type HelmParameter struct {
+	// Name is the name of the Helm parameter
+	Name string `json:"name,omitempty" mapstructure:"name,omitempty"`
+	// Value is the value for the Helm parameter
+	Value string `json:"value,omitempty" mapstructure:"value,omitempty"`
+}
+
+// RawSource holds raw specific options
+type RawSource struct{}
+
+// ExplicitType returns the type (e.g., Helm, Kustomize, etc.) of the application. If either none or multiple types are defined, returns an error.
+func (gitSource *GitSource) ExplicitType() (SourceType, error) {
+	var appTypes []SourceType
+	if gitSource.Kustomize != nil {
+		appTypes = append(appTypes, SourceTypeKustomize)
+	}
+	if gitSource.Helm != nil {
+		appTypes = append(appTypes, SourceTypeHelm)
+	}
+	if gitSource.Raw != nil {
+		appTypes = append(appTypes, SourceTypeRaw)
+	}
+	if len(appTypes) == 0 {
+		// Fallback to a raw source type if a user has not specified anything.
+		return SourceTypeRaw, nil
+	}
+	if len(appTypes) > 1 {
+		typeNames := make([]string, len(appTypes))
+		for i := range appTypes {
+			typeNames[i] = string(appTypes[i])
+		}
+		return "", fmt.Errorf("multiple sources defined: %s", strings.Join(typeNames, ","))
+	}
+	appType := appTypes[0]
+	return appType, nil
+}
+
+type RepoCredential struct {
+	URL            string          `json:"url" mapstructure:"url"`
+	HTTPCredential *HTTPCredential `json:"httpCredential,omitempty" mapstructure:"httpCredential,omitempty"`
+	SSHCredential  *SSHCredential  `json:"sshCredential,omitempty" mapstructure:"sshCredential,omitempty"`
+	TLS            *TLS            `json:"tls,omitempty" mapstructure:"tls,omitempty"`
```

Comment:
these didn't have omitempty before - which actually seemed to work okay - but I think we are technically supposed to have them for pointer types